core/arch/interp.c

```diff
@@ -817,6 +817,27 @@ check_new_page_jmp(dcontext_t *dcontext, build_bb_t *bb, app_pc new_pc)
 }
 
 static inline void
+bb_process_single_step(dcontext_t *dcontext, build_bb_t *bb)
+{
+    LOG(THREAD, LOG_INTERP, 2, "interp: single step exception bb at "PFX"\n", bb->instr_start);
+    instrlist_append(bb->ilist, bb->instr);
+    /* Sets exit reason dynamically. */
+    instrlist_preinsert(bb->ilist, bb->instr,
+                        instr_create_save_immed16_to_dcontext(dcontext, EXIT_REASON_SINGLE_STEP,
+                                                              EXIT_REASON_OFFSET));
+
+    /* Mark instruction as special exit. */
+    instr_branch_set_special_exit(bb->instr, true);
+    bb->exit_type |= LINK_SPECIAL_EXIT;
+
+    /* Make this bb thread-private and a trace barrier. */
+    bb->flags &= ~FRAG_SHARED;
+    bb->flags |= FRAG_CANNOT_BE_TRACE;
+    /* So that the fragment does not get shared. */
+    bb->flags |= FRAG_TEMP_PRIVATE;
```

Comment:
> We need to make this thread-private by remove FRAG_SHARED from bb->flags

That was more complicated. I did not find a better solution than making 
`bb->flags |= FRAG_TEMP_PRIVATE` 
for a limited time (see a few lines down) but I am sure there is a better solution.